File: apps/web/lib/api/financial-intelligence.ts
import { publicEnv } from "@/lib/config";

export type Insight = {
  title: string;
  severity: "info" | "warning" | "positive";
  message: string;
  metric_value: string | null;
};

export type FinancialSummary = {
  profit: {
    revenue: string;
    expenses: string;
    profit: string;
    profit_margin: string;
    previous_month_profit: string;
    profit_trend_percent: string | null;
  };
  cashflow: {
    cash_position: string;
    receivables: string;
    payables: string;
    net_cash_flow: string;
    cash_risk_level: string;
    cash_risk_reason: string;
  };
  gst: {
    gst_collected: string;
    gst_input_paid: string;
    current_gst_payable: string;
    estimated_month_end_liability: string;
  };
  insights: Insight[];
};

async function get<T>(path: string, token: string): Promise<T> {
  const response = await fetch(`${publicEnv.NEXT_PUBLIC_API_URL}${path}`, {
    headers: { Authorization: `Bearer ${token}` }
  });
  if (!response.ok) {
    const error = await response.json().catch(() => ({ detail: "Request failed" }));
    throw new Error(typeof error.detail === "string" ? error.detail : "Request failed");
  }
  return response.json() as Promise<T>;
}

export const financialIntelligenceApi = {
  summary: (companyId: string, token: string, month: string) =>
    get<FinancialSummary>(
      `/companies/${companyId}/financial-intelligence/summary?month=${month}-01`,
      token
    )
};

File: apps/api/app/schemas/financial_intelligence.py
from datetime import date
from decimal import Decimal

from pydantic import BaseModel


class IntelligencePeriod(BaseModel):
    month: date
    starts_on: date
    ends_on: date


class ProfitIntelligence(BaseModel):
    period: IntelligencePeriod
    revenue: Decimal
    expenses: Decimal
    profit: Decimal
    profit_margin: Decimal
    previous_month_profit: Decimal
    profit_trend_percent: Decimal | None


class CashFlowIntelligence(BaseModel):
    period: IntelligencePeriod
    cash_position: Decimal
    receivables: Decimal
    payables: Decimal
    net_cash_flow: Decimal
    cash_risk_level: str
    cash_risk_reason: str


class GstIntelligence(BaseModel):
    period: IntelligencePeriod
    gst_collected: Decimal
    gst_input_paid: Decimal
    current_gst_payable: Decimal
    estimated_month_end_liability: Decimal


class OwnerInsight(BaseModel):
    title: str
    severity: str
    message: str
    metric_value: Decimal | None = None


class FinancialIntelligenceSummary(BaseModel):
    profit: ProfitIntelligence
    cashflow: CashFlowIntelligence
    gst: GstIntelligence
    insights: list[OwnerInsight]

